lambda-buffers-codegen/src/LambdaBuffers/Codegen/Rust/Print/Derive.hs

```diff
@@ -0,0 +1,270 @@
+module LambdaBuffers.Codegen.Rust.Print.Derive (printDeriveEqBase, printDeriveIsPlutusData, printDeriveJson, rsTraitImplPrinters) where
+
+import Control.Lens ((^.))
+import Data.Foldable (for_)
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Set (Set)
+import Data.Set qualified as Set
+import LambdaBuffers.Codegen.LamVal.Eq (deriveEqImpl)
+import LambdaBuffers.Codegen.LamVal.Json (deriveFromJsonImpl, deriveToJsonImpl)
+import LambdaBuffers.Codegen.LamVal.MonadPrint qualified as LV
+import LambdaBuffers.Codegen.LamVal.PlutusData (deriveFromPlutusDataImpl, deriveToPlutusDataImpl)
+import LambdaBuffers.Codegen.Print qualified as Print
+import LambdaBuffers.Codegen.Rust.Print (MonadPrint)
+import LambdaBuffers.Codegen.Rust.Print.LamVal (printInstance)
+import LambdaBuffers.Codegen.Rust.Print.Syntax qualified as R
+import LambdaBuffers.ProtoCompat qualified as PC
+import Prettyprinter (Doc, align, braces, colon, comma, encloseSep, hcat, indent, lparen, parens, rparen, space, vsep, (<+>))
+import Proto.Codegen qualified as P
+import Proto.Codegen_Fields qualified as P
+
+rsTraitImplPrinters ::
+  MonadPrint m =>
+  Map
+    R.QTraitName
+    ( PC.ModuleName ->
+      PC.TyDefs ->
+      (Doc ann -> Doc ann) ->
+      PC.Ty ->
+      m (Doc ann)
+    )
+rsTraitImplPrinters =
+  Map.fromList
+    [
+      ( R.qLibRef R.MkTraitName "std" "cmp" "PartialEq"
+      , printDerivePartialEqBase
+      )
+    ,
+      ( R.qLibRef R.MkTraitName "std" "cmp" "Eq"
+      , printDeriveEqBase
+      )
+    ,
+      ( R.qLibRef R.MkTraitName "plutus-ledger-api" "plutus_data" "IsPlutusData"
+      , printDeriveIsPlutusData
+      )
+    ,
+      ( R.qLibRef R.MkTraitName "lbr-prelude" "json" "Json"
+      , printDeriveJson
+      )
+    ]
+eqTraitMethodName :: R.ValueName
+eqTraitMethodName = R.MkValueName "eq"
+
+eqTraitMethodArgs :: [(R.ValueName, R.QTyName)]
+eqTraitMethodArgs = [(R.MkValueName "self", R.qBuiltin R.MkTyName "Self"), (R.MkValueName "other", R.qBuiltin R.MkTyName "Self")]
+
+eqTraitMethodReturns :: R.QTyName
+eqTraitMethodReturns = R.qBuiltin R.MkTyName "bool"
+
+lvEqBuiltinsBase :: LV.PrintRead R.QValName
+lvEqBuiltinsBase = LV.MkPrintRead $ \(_ty, refName) ->
+  Map.lookup refName $
+    Map.fromList
+      [ ("eq", R.qLibRef R.MkValueName "lbr-prelude" "lamval" "eq")
+      , ("and", R.qLibRef R.MkValueName "lbr-prelude" "lamval" "and")
+      , ("true", R.qBuiltin R.MkValueName "true")
+      , ("false", R.qBuiltin R.MkValueName "false")
+      , ("PhantomData", R.qLibRef R.MkValueName "std" "marker" "PhantomData")
```

Comment:
This seems fishy. PhantomData is not part of the LamVal vocabulary for Eq. I wonder why it's here?

Reply:
fixed